receiver/flinkmetricsreceiver/integration_test.go

```diff
@@ -0,0 +1,203 @@
+	ws = waitStrategy{endpoint: "http://localhost:8081/jobs/metrics"}
+	err = ws.waitFor(context.Background(), "")
+	require.NoError(t, err)
+
+	// Override function to return deterministic field
+	defer func() { osHostname = os.Hostname }()
+	osHostname = func() (string, error) { return "job-localhost", nil }
+
+	// Override function to return deterministic field
+	defer func() { taskmanagerHost = strings.Split }()
+	taskmanagerHost = func(id string, sep string) []string { return []string{"taskmanager-localhost"} }
+
+	// Override function to return deterministic field
+	defer func() { taskmanagerID = reflect }()
+	taskmanagerID = func(id string) string { return "taskmanagerID" }
+
+	f := NewFactory()
+	cfg := f.CreateDefaultConfig().(*Config)
+	cfg.ScraperControllerSettings.CollectionInterval = 100 * time.Millisecond
+	cfg.Endpoint = fmt.Sprintf("http://%s", net.JoinHostPort(hostname, "8081"))
+
+	consumer := new(consumertest.MetricsSink)
+	settings := componenttest.NewNopReceiverCreateSettings()
+	rcvr, err := f.CreateMetricsReceiver(context.Background(), settings, cfg, consumer)
+	require.NoError(t, err, "failed creating metrics receiver")
+	require.NoError(t, rcvr.Start(context.Background(), componenttest.NewNopHost()))
+	require.Eventuallyf(t, func() bool {
+		return consumer.DataPointCount() > 0
+	}, 2*time.Minute, 1*time.Second, "failed to receive more than 0 metrics")
+	require.NoError(t, rcvr.Shutdown(context.Background()))
+
+	actualMetrics := consumer.AllMetrics()[0]
+	expectedFile := filepath.Join("testdata", "integration", "expected.json")
+	expectedMetrics, err := golden.ReadMetrics(expectedFile)
+	require.NoError(t, err)
+	require.NoError(t, scrapertest.CompareMetrics(expectedMetrics, actualMetrics, scrapertest.IgnoreMetricValues()))
+}
+
+func getContainer(t *testing.T, req testcontainers.ContainerRequest) testcontainers.Container {
+	require.NoError(t, req.Validate())
+	container, err := testcontainers.GenericContainer(
+		context.Background(),
+		testcontainers.GenericContainerRequest{
+			ContainerRequest: req,
+			Started:          true,
+		})
+	require.NoError(t, err)
+	return container
+}
+
+type waitStrategy struct {
+	endpoint string
+}
+
+func (ws waitStrategy) WaitUntilReady(ctx context.Context, st wait.StrategyTarget) error {
+	if err := wait.ForListeningPort("8081").
+		WithStartupTimeout(2*time.Minute).
+		WaitUntilReady(ctx, st); err != nil {
+		return err
+	}
+
+	hostname, err := st.Host(ctx)
+	if err != nil {
+		return err
+	}
+
+	return ws.waitFor(ctx, hostname)
+}
+
+// waitFor waits until an endpoint is ready with an id response.
+func (ws waitStrategy) waitFor(ctx context.Context, _ string) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(5 * time.Second):
+			return fmt.Errorf("server startup problem")
+		case <-time.After(100 * time.Millisecond):
+			resp, err := http.Get(ws.endpoint)
+			if err != nil {
+				continue
+			}
+
```

Comment:
I think timers would be a good fit for this

```suggestion
        ticker := time.NewTicker(100 * time.Millisecond)
        defer ticker.Stop()
        timeout := time.After(5 * time.Second)

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-timeout:
			return fmt.Errorf("server startup problem")
		case <-ticker.C:
			resp, err := http.Get(ws.endpoint)
			if err != nil {
				continue
			}

```

Reply:
updated